ref/vk/vk_devmem.c

```diff
@@ -1,43 +1,174 @@
+
+	vk_devmem_allocation_stats_t *const stats = &g_devmem.stats[type];
+	
+	/* Update allocations stats. */
+	
+	// Update current allocations.
+	stats->current.allocations += 1;
+	stats->current.allocated   += size;
+
+	// Update peak allocations.
+	if ( stats->peak.allocations < stats->current.allocations )
+		stats->peak.allocations = stats->current.allocations;
+
+	if ( stats->peak.allocated < stats->current.allocated )
+		stats->peak.allocated = stats->current.allocated;
+
+	/* Update alignment holes stats. */
+
+	int alignment_hole = size % alignment;
+	if ( alignment_hole > 0 ) {
+		// Update current alignment holes stats.
+		stats->current.align_holes      += 1;
+		stats->current.align_holes_size += alignment_hole;
+
+		// Update peak alignment holes stats.
+		if ( stats->peak.align_holes < stats->current.align_holes )
+			stats->peak.align_holes = stats->current.align_holes;
+
+		if ( stats->peak.align_holes_size < stats->current.align_holes_size )
+			stats->peak.align_holes_size = stats->current.align_holes_size;
+
+		if ( stats->peak.align_hole_size < alignment_hole )
+			stats->peak.align_hole_size = alignment_hole;
+	}
+}
+
+// Register deallocation (freeing) in stats of the provided type.
+static void register_free_for_type( vk_devmem_usage_type_t type, int size, int alignment ) {
+	ASSERT( type >= VK_DEVMEM_USAGE_TYPE_ALL );
+	ASSERT( type <  VK_DEVMEM_USAGE_TYPES_COUNT );
+
+	vk_devmem_allocation_stats_t *const stats = &g_devmem.stats[type];
+
+	/* Update current allocations stats. */
+
+	stats->current.allocations -= 1;
+	stats->current.allocated   -= size;
+	
+	/* Update current alignment holes stats. */
+
+	int alignment_hole = size % alignment;
+	if ( alignment_hole > 0 ) {
+		stats->current.align_holes      -= 1;
+		stats->current.align_holes_size -= size;
+	}
+}
+
+#define VKMEMPROPFLAGS_COUNT 5
+#define VKMEMPROPFLAGS_STRLEN (VKMEMPROPFLAGS_COUNT + 1)
+
+// Fills string `out_flags` with characters at each corresponding flag slot.
+static void VK_MemoryPropertyFlags_String( VkMemoryPropertyFlags flags, char out_flags[VKMEMPROPFLAGS_STRLEN] ) {
+	int flag = 0;
+	if ( flags & VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT     )  {out_flags[flag] = 'D';}  else  {out_flags[flag] = '-';}  flag += 1;
+	if ( flags & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT     )  {out_flags[flag] = 'V';}  else  {out_flags[flag] = '-';}  flag += 1;
+	if ( flags & VK_MEMORY_PROPERTY_HOST_COHERENT_BIT    )  {out_flags[flag] = 'C';}  else  {out_flags[flag] = '-';}  flag += 1;
+	if ( flags & VK_MEMORY_PROPERTY_HOST_CACHED_BIT      )  {out_flags[flag] = '$';}  else  {out_flags[flag] = '-';}  flag += 1;
+	if ( flags & VK_MEMORY_PROPERTY_LAZILY_ALLOCATED_BIT )  {out_flags[flag] = 'L';}  else  {out_flags[flag] = '-';}  flag += 1;
+	// VK_MEMORY_PROPERTY_PROTECTED_BIT
+	// VK_MEMORY_PROPERTY_DEVICE_COHERENT_BIT_AMD
+	// VK_MEMORY_PROPERTY_DEVICE_UNCACHED_BIT_AMD
+	// VK_MEMORY_PROPERTY_RDMA_CAPABLE_BIT_NV
+	out_flags[flag] = '\0';
```

Comment:
(абсолютно опционально)
```suggestion
        out_flags[0] = (flags & VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT) ? 'D' : '-';
        out_flags[1] = (flags & VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT) ? 'V' : '-';
        ...
	out_flags[5] = '\0';
```

Reply:
Это я кардинально поменял, как ты и предлагал, через `PRI_xxxFLAGS_FMT` и `PRI_xxxFLAGS_ARG(...)`.